[ENH] Add Differentiable Transformer and Exact PDFs for TargetTransformRegressor

joshdunnlime opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://skpro.readthedocs.io/en/latest/contribute.html
-->

#### Reference Issues/PRs
Background on why we want this: #601 and #605
Supersedes this PR: #608

#### What does this implement/fix? Explain your changes.
This implementation adds a DifferentiableTransformer (DT). This acts as a wrapped around an sklearn transformer and gives the user the option to use either (in order):
1) an explicit derivative via `inverse_func_diff`.
2) an explicit derivative for a scaler transformer via `scaler_`.
3) Numerical differentiation of the `inverse_transform`.

The derivative available for both the (forward) transform and inverse_transform.

The DT has a coerce classmethod that takes a sklearn transformer or a function and coerces it to a DT.

In addition to the above, it also makes changes to #605 to apply the Jacobian (derivative of the transform/inverse_tranform) to the `pdf` and `log_pdf`. In cases 1) and 2) above this returns the exact pdf and log_pdf.

This preserves the current functionality of being able to pass a function to TTR (TransformedTargetRegressor) and allows a user to extend this by passing an sklearn transformer or their own DT. The final point allows the user to configure their DT with either the explicit derivative, or to pass kwargs that can be used for numerical differentiation. In theory, it also allows a user to pass these kwargs as hyperparameters via a gridsearch/optimisation.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

1) Implementation of the DT is the most important part:

   - Are we happy with the transformer inheriting from sklearn?
   - Do we need this much abstraction at this stage?
   - Do we wish to add/implement tags for this?

2) Changes to TD (TranformedDistribution) - the main change here is internally handling a transformer instead of a function.
3) Changes to the TTR are minor.

On points 2) and 3), note that if we wish to explicitly pass `transform` only as a method/function, it is very straightforward to change this back. We would simply need to add `inverse_transform` and `inverse_func_diff` as kwargs to the TD. This does mean adding more kwargs if we wish to support numerical derivative kwargs.

Secondary feedback 

#### Did you add any tests for the change?

Yes, added a param3 to TD, passing a FunctionTransformer instead of just a function.

#### Any other comments?

I had the bulk of this implementation down prior to the discussion on what we pass to TTR/TD [here](https://github.com/sktime/skpro/issues/601#issuecomment-3362967573). Again, it is trivial to move the creation of the DT into TTR and pass the `transform`, `inverse_transform` and `inverse_func_diff`, but I feel the current approach is much cleaner as it keep the majority of the new logic in the DT and within the new _transformer.py module.

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).


<!--
Thanks for contributing!
-->

joshdunnlime committed:
feat: replace inverse with whole transformer on transformed distribution.

This allows us to get more information from the tranformer class. E.g. MinMaxScaler scale_ parameter.

Though not part of the typical forward-facing interface, TransformedDistribution.transform is now no longer callable. Instead TransformedDistribution.transform.inverse_transform would be needed.

joshdunnlime committed:
feat: allows `n_trails=0` for fast optimisation free training

This is useful when testing XGBoostLLS and hyperparameter optimisation is not important. XGBoostLLS can be computationally expensive when searching over the default 30 trials.

joshdunnlime committed:
Merge branch 'xgboostlss-no-trails' into ttr-logloss

joshdunnlime committed:
feat: adds log_pdf to the TargetTranformRregressor

Does so by adding the log_pdf with change-of-variables (jacobian of inverse transform) and direct method for linear transforms, and numerical method for non-linear transforms.

TransformedDistribution now take transformer instead of transform - that is, the transformer class instead of just the inverse transform function/method.

joshdunnlime committed:
docs: Add notebook to show jacobian calcs

This is not actually intended as an example. It is a simple way to share findings. This can (should) be removed before any final merges into main.

joshdunnlime committed:
test: fix TransformdDistribution test with np.exp as inverse_func

joshdunnlime committed:
Revert "Merge branch 'xgboostlss-no-trails' into ttr-logloss"

This reverts commit 4146a6f87a2bd645abc6164fa015159050683406, reversing
changes made to 8bc58970ed09fba334f323971226654b45d0beb1.

joshdunnlime committed:
docs: fix doctest example for TransformedDistribution

joshdunnlime committed:
feat: allows use of user defined `tranform_diff` function on an skpro transformer

joshdunnlime committed:
feat: add DifferentiableTransformer and base classes

joshdunnlime committed:
feat: Implement more of the sklearn -> skpro transformer logic

joshdunnlime committed:
add pdf and keep log_pdf

joshdunnlime committed:
remove sample notebook

joshdunnlime committed:
add pdf

joshdunnlime committed:
feat: add approximation of the pdf and log_pdf functions

joshdunnlime committed:
feat: add the _pdf and _log_pdf methods with index and column info

joshdunnlime committed:
fix: change TranformedDistribution measuretype to mixed

joshdunnlime committed:
feat: add index and column information to internal distribution

This is needed so that when TransformerDistribtution.distribution is called, it has the same indices as the wrapper distribution.

joshdunnlime committed:
fix: add pdf and log_pdf to approx to pass pdf_log_pdf test

joshdunnlime committed:
feat: add the DifferentiableTransformer and parent classes

joshdunnlime committed:
replace transform function with DifferentiableTransformer class

joshdunnlime committed:
use the distribution pdf and log_pdf method (remove scipy workaround)

joshdunnlime committed:
Merge branch 'ttr-logloss' into differentiable-transformer

joshdunnlime committed:
reinstate try-except in fit_with_fitted

joshdunnlime commented:
Usage examples:

Create some data:

```python
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

from skpro.metrics import LogLoss, CRPS
from skpro.regression.xgboostlss import XGBoostLSS
from skpro.regression.compose import TransformedTargetRegressor
from skpro.compose import DifferentiableTransformer

size = 1000

X = np.random.normal(1, 1, size)
y = (2 * X + np.random.normal(0, 0.5, size)) # / 15 + 0.5 # for Beta

Xy = pd.DataFrame(
    {"target": y, "feature": X},
    index=range(size)
)
```

Now we can create pass an sklearn transform to TTR:

```python
xgb = XGBoostLSS(**params)
mms = MinMaxScaler((0.1, 0.9))
pipe = TransformedTargetRegressor(regressor=xgb, transformer=mms)
```

or a DT:

```python
xgb = XGBoostLSS(**params)
mms = MinMaxScaler((0.1, 0.9))
mms = DifferentiableTransformer(transformer=MinMaxScaler())
pipe = TransformedTargetRegressor(regressor=xgb, transformer=mms)
```

Both are handled the same:

```python
pipe.fit(X=Xy[["feature"]], y=Xy["target"])
pp = pipe.predict_proba(Xy[["feature"]])
p = pipe.predict(Xy[["feature"]])
CRPS()(y_true=Xy["target"], y_pred=pp)
LogLoss()(y_true=Xy["target"], y_pred=pp)
```
> 0.41678

And in either case, if we inspect the TD (`pp`) returned by `predict_proba` we get a DT as the transformer.

```python
pp.transformer_
```
> DifferentiableTransformer(transformer=MinMaxScaler(feature_range=(0.1, 0.9)))

joshdunnlime committed:
add inverse function to init

joshdunnlime committed:
allow exact cdf

joshdunnlime committed:
default approx (assuming no inverse)

joshdunnlime committed:
Merge branch 'main' into differentiable-transformer

joshdunnlime committed:
Merge branch 'main' into differentiable-transformer

fkiraly committed:
Merge branch 'main' into pr/612

fkiraly commented:
I tried to resolve the conflicts, have a look

joshdunnlime committed:
feat: add simple and numdifftools differentiation

joshdunnlime committed:
Merge branch 'differentiable-transformer' of https://github.com/joshdunnlime/skpro into differentiable-transformer

joshdunnlime committed:
improve approx and exact tags and indices None handling

joshdunnlime committed:
feat: add correct tag logic to manage exact and approx scenarios

Here we consider the numerical differentiation to classed as approx. For wellbehaved function (which change-of-variables requires), this is likely to be overkill.

fkiraly commented:
ok, I also think that transformer is the better way to go after some thinking.

How do we proceed practically - do you want to give it a stab and then we maybe refine?

joshdunnlime committed:
fix: refactor to remove numdifftools and use scipy derviative with fallback

joshdunnlime committed:
Merge branch 'main' into differentiable-transformer

joshdunnlime committed:
fix: fix new scipy derivative usage

fkiraly commented:
Nice! Will review in the coming days - this is much appreciated, but I think I need more time to digest the API design.
(please ping if I need more than a few days)

fkiraly reviewed (changes requested):
Great!

Two remarks on the design:

* can you make sure that `TransformedDistribution` does not deviate from the current logic in the function case? The current design causes default approximations for the pdf etc to live in two places.
* I think your transformation design is not quite there yet:
    * it should work if `transform` is called on a different `X` than `fit`
    * I do not get the index saving and restoring logic in `fit` and `transform`. Why is this necessary? Should the index not be the same as `X` passed in `transform`, and the columns determined by what the transformation does?

joshdunnlime committed:
refactor: refactor transformer classes to be much more modular and composable

Adds a simple BaseTransformer implementation. Extend this with a WrapTransformer and BaseDifferentiableTransformer to allow the addition of differentiable transformers that do not rely on wrapping sklearn transformers.

joshdunnlime committed:
fix: add __init__.py for test dirs

joshdunnlime committed:
refactor: improve TTR inner transform vs sklearn transform nomenclature

Also adds improved tests, directly for PDFs and better docstrings

joshdunnlime committed:
fix: fix switch of transform and inverse_transform with coerce_diff

also fixes has_transform_func fix to catch missing inner_inverse_transform

joshdunnlime committed:
test: test refitting and transform index logic

joshdunnlime commented:
@fkiraly - are you happy with the new structure? If so, I happy to improve docs and add examples, ready for merging. Thanks.

fkiraly commented on skpro/distributions/trafo/_transformed.py:
```diff
@@ -18,26 +23,50 @@ class TransformedDistribution(BaseDistribution):
     * ``transform``: a function that is applied to the distribution.
       This can be entry-wise, or a ``pandas.DataFrame`` to ``pandas.DataFrame``
       function that can be applied to samples from ``distribution``.
+      NOTE: this is the transform that is applied to the distribution, not
```
I do not quite get the note. I think this is too confusing, instead I would add more details on what the function is in terms of random variables. It is simply the distribution of f(X) if X is a random variable distributed according to `distribution`, and f=`transform`, no?

I think the docstring should be very explicit what the distribution this represents is, in all cases of what `transform` can be.

joshdunnlime replied:
Agreed. This could be clearer and should be explicit about all potential cases of `transform`.
Theoretically, the RV framing is useful - I will add that.

fkiraly reviewed (changes requested):
I have a few questions for my understanding before I continue the review:

* how are we solving the problem that the `TransformedDistribution` resets the fitted `transformer_` upon passing from `TransformedTargetRegressor`? Previously, we were circumventing this by passing the methods of `self.transformer_` instead of the object.
* Did you, or not, flip around the convention where the `inverse_transform` from the `self.transformer_` in `TransformedTargetRegressor` gets passed as `transform` to `TransformedDsitribution`? There are also two steps where it could flip around after this PR: once when passing it to the `DifferentiableTransformer`, and once when the `DifferentiableTransformer` is applied within the `TransformedDistribution`. Can you explain what your intention is, whether it has changed or not compared to prior, and why you think what you are doing is the right design?

Regarding the design:

* I would wrap in `DifferentiableTransformer` *only*  if we know an exact derivative is available.
* I think there are too many case distinctions and non-DRY duplications throughout - can you please look into this? Is it possible to use, instead of case distinctions, the pattern of coercions? E.g., case distinctions based on `scipy`, why not wrap them in a facade pattern, etc.

joshdunnlime commented:
> how are we solving the problem that the TransformedDistribution resets the fitted transformer_ upon passing from TransformedTargetRegressor? Previously, we were circumventing this by passing the methods of self.transformer_ instead of the object.

Here we use the `_fit_with_fitted` method. If the transformer is fitted it returns the original instance without re-fitting. You suggested looking into using the `FrozenEstimator` so I will do that but it would be worth aligning on the coercion side of things first as that seems a much more fundamental decision.

fkiraly commented:
> A "double-flip" only occurs if someone passed a func to transform (which is coerced to an FT).

I do not think this should happen. I think there should only ever be a single flip. Since there are two, something in the design must be problematic, though I cannot pin it down.

Can you please explain, again, very explicitly, in which situation there is a double-flip? Optimally with pseudo-code, since I feel it will be very hard to follow if just described in English.

> The idea of the DT is to use the facade pattern (your second point). By having the DT we can use explicit differentiation when we have it, otherwise, numerical differentiation is used and the PDF, log-PDF, energy-score and CRPS can all be computed. If we do not coerce to a DT then we need to handle more cases distinctions closer to the user.

But we are still duplicating even if we do that; the logic for numerical differentiation that gets called moves from the distribution default `_pdf` to the DT. I think that introduces more case distinctions on the logic in `_pdf` that can be called, instead of reducing their number?

Please correct me if you think I'm wrong.

> so I will do that but it would be worth aligning on the coercion side of things first as that seems a much more fundamental decision.

Can you please write up the exact cases that we would be handling?

I think the cases should be as follows:

**in TD**, on `transform`:

* if function, applied directly; `inverse_transform` can also be passed
* if `sklearn` transformer, applied (fitted) via `transform` and `inverse_transform`
* if DT, its `transform` , `transform_diff`, and `inverse_transform` are used

**in TTR**:

no case distinctions, fitted transform is passed through directly to TD

joshdunnlime commented:
> I do not think this should happen. I think there should only ever be a single flip. Since there are two, something in the design must be problematic, though I cannot pin it down.

I would say the issue is just that the `TransformedDistribution` uses the term `transform` to mean the backward transform: `g(y_t) -> y`; and uses the term `inverse_transform` to mean the forward transform: `f(y) -> y_t`, where `y` and `y_t` are the original target and transformed target. This is contrary to the standard sklearn nomenclature.

> Can you please explain, again, very explicitly, in which situation there is a double-flip? Optimally with pseudo-code, since I feel it will be very hard to follow if just described in English.

Yes - this happens in only one occasion:
```python
td = TransformedDistribution(transform=fx)
```

where `fx` is some callable AND `fx` is not a `DifferentiableTransformer` AND `fx` is not a `Transformer`. E.g.

```python
assert callable(fx)  # True
assert not isinstance(transform, TransformerMixin)  # True
assert not isinstance(transform, DifferentiableTransformer)  # True
```

In this case only, the first flip happens in the `_coerce_to_diff_transformer` as `transform` is passed to the [FunctionTransformer](https://github.com/joshdunnlime/skpro/blob/07df29d583a8e0b2e50d7f26a509c77c3e7db796/skpro/distributions/trafo/_transformed.py#L617) as `inverse_func=transform` (and `func=inverse_transform`).

In all other cases, the passed transformer (Diff or Sklearn) is just directly [passed through](https://github.com/joshdunnlime/skpro/blob/07df29d583a8e0b2e50d7f26a509c77c3e7db796/skpro/distributions/trafo/_transformed.py#L595) or [lightly wrapped](https://github.com/joshdunnlime/skpro/blob/07df29d583a8e0b2e50d7f26a509c77c3e7db796/skpro/distributions/trafo/_transformed.py#L597).

Any other "flips" are done internally of the TD. In the callable (`fx`) case, this is the second flip. In the passed Transformer case, this will be the only flip. An example of this happening in the TD is [here](https://github.com/joshdunnlime/skpro/blob/07df29d583a8e0b2e50d7f26a509c77c3e7db796/skpro/distributions/trafo/_transformed.py#L375).

IMO, the "cleanest" solution would be to deprecate the passing of functions to TD and only pass transformers. This would mean at most, there was a single flip (internally to TD) and this private API could be inverted so no flips are done. Of course, the issue here is back-compat!

joshdunnlime committed:
test: improve tests

joshdunnlime committed:
feat: add func and inverse_func to ttr to conform with sklearn

joshdunnlime committed:
refactor: revert transform distirbution internal transform to D -> T for the forward transform/func

joshdunnlime committed:
Merge branch 'main' into differentiable-transformer

joshdunnlime committed:
fix: fix tag check logic

fkiraly reviewed (changes requested):
Nice!
 
* I think the API is a bit confused now - why do we nee `transformer` * and* `inverse_transformer`, if a transformer has both? I do not think the second, new argument makes sense, but happy to be convinced otherwise.
* I would suggest to leave only the `transform` argument for now, and allow it to be a differentiable transformer. And add potential additional arguments and features in separate PR. This is to keep complexity and number of iterations low.
* the try/except problem still needs to be addressed

joshdunnlime committed:
feat: add `_from_inverse_transform` classmethod

this is instead of `inverse_transform` kwarg which polutes the public init API

joshdunnlime commented:
An alternative to `inverse_transform` kwarg would be to add a classmethod to TD called `_from_inverse_transform`. This avoids polluting the public API and has all of the benefits listed above plus is much more explicit.

joshdunnlime committed:
feat: use _check_soft_deps and remove deprecation

joshdunnlime committed:
feat: remove base diff transformer class and simplify

joshdunnlime committed:
fix: fix doctests

joshdunnlime committed:
fix: fix doctest remove func kwarg and add import

joshdunnlime committed:
fix: fix doctest derivative output

joshdunnlime committed:
fix: add nicer derivative example without warning

joshdunnlime committed:
fix: prevent output of fit in doctest

joshdunnlime committed:
fix: fix scipy derivative imports

scipy 1.15.0 and up use differentiate, 1.12.0 and lower use misc, in between these two, there is no derivative function

joshdunnlime commented:
@fkiraly Any more thoughts on this one? Thanks!

joshdunnlime commented:
@fkiraly - this should be complete now. Would love to get this merged if you're happy with it.
